sdk/core/core-lro/docs/LROEngine.md

```diff
@@ -0,0 +1,167 @@
+
+The design consists of three main pieces:
+
+- an interface, named `LongRunningOperation<T>` which groups various primitives needed to implement LROs
+- a class, named `LroEngine`, that implements the LRO engine and its constructor takes as input an object that implements `LongRunningOperation<T>`
+- classes that implement `LongRunningOperation<T>` for `@azure/core-http` and `@azure/core-client`. @joheredi also created one for `@azure-rest/core-client` in https://github.com/Azure/azure-sdk-for-js/pull/15898
+
+### `LongRunningOperation<T>`
+
+This interface contains two methods: **sendInitialRequest** and **sendPollRequest**. I propose to make this interface exported by `@azure/core-lro`.
+
+#### `sendInitialRequest`
+
+This method should be implemented to send the initial request to start the operation and it has the following signature:
+
+```ts
+sendInitialRequest: (initializeState: (rawResponse: RawResponse, flatResponse: unknown) => boolean) => Promise<LroResponse<T>>
+```
+
+The `initializeState` parameter is a function that will initialize the state of the polling operation and will return a Boolean that indicates whether the operation has already been completed. This Boolean is especially useful in `@azure/core-client`'s engine because at this point, the customer-provided callback should be called. The implementation for `@azure/core-http` looks like this:
+
+```ts
+public async sendInitialRequest(
+  initializeState: (
+    rawResponse: RawResponse,
+    flatResponse: unknown
+  ) => boolean
+): Promise<LroResponse<T>> {
+  const response = await this.sendOperation(this.args, this.spec); // the class will have sendOperation, args, and spec as private fields
+  initializeState(response.rawResponse, response.flatResponse);
+  return response;
+}
+```
+
+#### `sendPollRequest`
+
+This method should be implemented to send a polling (GET) request, a request the service should respond to with the current status of the operation, and it has the following signature:
+
+```ts
+sendPollRequest: (path: string, isDone: (response: LroResponse<T>) => boolean) => Promise<LroResponse<T>>;
+```
+
+This method takes two things as input, the request path as you would expect but also a predicate `isDone` that can tell you whether the operation has finished. Similar to the `initializeState` in `sendInitialRequest`, this predicate is needed in clients that use `@azure/core-client` because the customer-provided callback will have to be called if the predicate returns true. `isDone` is implemented as a memoized predicate so that if the customer has already called it, `LroEngine` can just use the memoized result instead of recomputing it. Here is what a simplified implementation would look like for `@azure/core-http`:
```

Comment:
I'm a bit confused by the `isDone` param, since I can't seem to find meaningful usage of it anywhere in this PR. What scenarios is it needed for and can we add some tests for these?

Reply:
It is used in the auto-generated `coreClientLro.ts`: https://github.com/deyaaeldeen/autorest.typescript/blob/8e1ab5b7575bb0aaaf1d02ee93a0be0eff04807f/src/coreClientLro.ts#L43. This works by the fact that the integration tests in code gen pass since all assertion logic is done inside those callbacks. Here is the PR: https://github.com/Azure/autorest.typescript/pull/1099

Reply:
Why do we only call `onResponse` when it's done? The point of `onResponse` is to debug *every* HTTP call, not the last one.

Reply:
Ok I removed the `isDone` predicate param.